core/aws-lsp-fqn/src/node/runner.ts

```diff
@@ -0,0 +1,71 @@
+import { Worker } from 'worker_threads'
+import { FqnExtractor, FqnExtractorInput, FqnExtractorOutput, Result, RunnerConfig } from '../common/types'
+import path = require('path')
+
+export const fqnExtractor: FqnExtractor = (input: FqnExtractorInput, config?: RunnerConfig) => {
+    const { logger, timeout = 5000 } = config ?? {}
+
+    let hasFulfilled = false
+
+    let timer: NodeJS.Timeout
+    const worker = new Worker(path.resolve(__dirname, './fqnExtractorWorker.js'), {
```

Comment:
Do we have to create the worker thread on every invocation? Does performance of the WASM module degrade over time or instantly on malformed inputs? Does it hang or stall, or saturate the CPU?

I worry about the performance overhead given this is on the chat request path. The code itself doesn't seem to have much initialization overhead (it's just `extract`) but there is overhead is spawning the thread.

And `extract` is an `async` method so that implies to me there is some I/O wait or otherwise non-CPU bound processing happening? Could one worker thread handle multiple concurrent extractions?

Reply:
Second question: you have a `browser/runner` as well that's currently inactive. Do you plan to move some of the `Promise` logic to some reusable place when that gets implemented? Or do you expect the async and timeout behavior will be different across platforms?

Reply:
Indeed, using something like `workerpool` could improve here.

Reply:
Using `workerpool` now, which simplified some of the things!

Reply:
I think so far I have only experience the module crashing instantly. I haven't experience the stalling issue so that needs to be investigated. I just implemented this with the worst case assumption that the I/O will be blocked.

Using `workerpool` as Oksana suggested will reduce the overhead, by keeping workers initialized. It also manages the task queue for us, which I wasn't sure how to do.

For the second question, the two main things that differ between environment are 1) how the worker script is bundled and 2) how the fqn library is imported. The newest refactor has reduced the number custom config requried.